Docs: Functions referencing big GIT repositories cause FC crash (#13987)
kyma-project/kyma: docs/04-operation-guides/troubleshooting/svls-05-serverless-periodically-restarting.md

---
title: Serverless periodically restarting
---

## Symptom

> Serverless restarting every 10 minutes when reconciling git-sourced functions.

Reviewer: ```suggestion
Serverless restarting every 10 minutes when reconciling git-sourced Functions.
```
Assuming we mean the Kubernetes resource here, which should be capitalized.

## Cause

> Function controller is polling for changes in referenced git repositories. If you have a lot of git-sourced functions and they were deployed together in approximately the same time their git sources will be checked-out in a synchronised pulse (every 10 minutes). If you happen to reference large repositories (multi-repositories) there will be rhythmical high demand on cpu and I/O resources necessary to checkout repositories that may cause function controller to crash and restart.

Reviewer: ```suggestion
Function controller is polling for changes in referenced git repositories. If you have a lot of git-sourced Functions and they were deployed together at approximately the same time, their git sources will be checked out in a synchronised pulse (every 10 minutes). If you happen to reference large repositories (multi-repositories), there will be rhythmical high demand on cpu and I/O resources necessary to checkout repositories. This may cause Function controller to crash and restart.
```

## Remedy

> Avoid using multi-repositories or large repositories to source your git functions. Using small or dedicated function repositories will decrease cpu and I/O  resources used to checkout git sources and hence assure better stability of functions controller.

Reviewer: ```suggestion
Avoid using multi-repositories or large repositories to source your git Functions. Using small, dedicated Function repositories decreases the cpu and I/O resources used to check out git sources, and hence improve the stability of Function controller.
```
Is it really "or"? I'm assuming a dedicated Function repo should still be small?

---
title: Serverless periodically restarting
---

## Symptom

> Serverless restarting every 10 minutes when checking for function changes.

Reviewer 2: ```suggestion
Serverless restarting every 10 minutes when reconciling git-sourced functions.
```

## Cause

> Git functions referencing large repositories cost a lot of resources to pull repositories and check for function changes. That causes serverless to restart.

Reviewer 2: ```suggestion
Function controller is polling for changes in referenced git repositories. If you have a lot of git-sourced functions and they were deployed together in approximately the same time their git sources will be checked-out in a synchronised pulse (every 10 minutes). If you happen to reference large repositories (multi-repositories) there will be rhythmical high demand on cpu and I/O resources necessary to checkout repositories that may cause function controller to crash and restart.
```

## Remedy

> Avoid using multi-repositories or large repositories to source your git functions. Using small or dedicated function repositories will decrease resources used and prevent serverless from restarting.

Reviewer 2: ```suggestion
Avoid using multi-repositories or large repositories to source your git functions. Using small or dedicated function repositories will decrease cpu and I/O  resources used to checkout git sources and hence assure better stability of functions controller.
```
